src/crowdin_client.cpp

```diff
@@ -143,139 +158,331 @@ dispatch::future<void> CrowdinClient::Authenticate()
+dispatch::future<void> CrowdinClient::DownloadFile(const long project_id,
                                                    const Language& lang,
+                                                   const long file_id,
+                                                   const std::string& file_extension,
                                                    const std::wstring& output_file)
 {
-    // NB: "export_translated_only" means the translation is not filled with the source string
-    //     if there's no translation, i.e. what Poedit wants.
-    auto url = "/api/project/" + project_id + "/export-file"
-                   "?json="
-                   "&export_translated_only=1"
-                   "&language=" + lang.LanguageTag() +
-                   "&file=" + http_client::url_encode(file);
-    return m_api->download(url, output_file);
+    wxLogTrace("poedit.crowdin", "DownloadFile(project_id=%ld, lang=%s, file_id=%ld, file_extension=%s, output_file=%S)", project_id, lang.LanguageTag(), file_id, file_extension.c_str(), output_file.c_str());
+    wxString ext(file_extension);
+    ext.MakeLower();
+    return m_api->post(
+        "projects/" + std::to_string(project_id) + "/translations/builds/files/" + std::to_string(file_id),
+        json_data({
+            { "targetLanguageId", lang.LanguageTag() },
+            // for XLIFF and PO files should be exported "as is" so set to `false`
+            { "exportAsXliff", !(ext == "xliff" || ext == "xlf" || ext == "po") }
+        }))
+        .then([this, output_file] (json r) {
+            std::string url = r["data"]["url"];
+            wxURI uri(url);
+            // Per download (local) client must be created since different domain
+            // per request is not allowed by HTTP client backend on some platorms.
+            // (e.g. on Linux).
+            auto downloader = std::make_shared<crowdin_http_client>(*this, std::string((uri.GetScheme() + "://" + uri.GetServer()).mb_str()));
+            wxLogTrace("poedit.crowdin", "Gotten file URL: %s", r.dump().c_str());
+            // Below capturing of `[downloader]` is needed to preserve `downloader` object
+            // from being destroyed before `download(...)` completes asynchroneously
+            // (what happens already after current function returns)
+            return downloader->download(url, output_file).then([downloader]() {});
+        });
 }
 
 
-dispatch::future<void> CrowdinClient::UploadFile(const std::string& project_id,
-                                                 const std::wstring& file,
+dispatch::future<void> CrowdinClient::UploadFile(const long project_id,
                                                  const Language& lang,
+                                                 const long file_id,
+                                                 const std::string& file_extension,
                                                  const std::string& file_content)
 {
```

Comment:
The 'const' for numerical arguments is superfluous (here and elsewhere: numbers are passed by value in C++; const& is needed when passing by reference to avoid modification).

Reply:
I just like this `const` C++ feature exactly here not to prevent modifying of external (outside/input) variable passed as argument to function (I am well aware about passing by `value` vs `reference`) but _to avoid just occasional unattended modifying of such arguments **inside function locally**_ (with e.g `=` or passing it as ref to some other function). It makes me more calm and sure (and BTW Rust uses such approach by default out of the box) about local code inside function since add one more compile-time check to avoid at least fool-proof and just occasional (by distraction) changes. While it does not cost too much redundancy/verbosity to the code. But anyway if you dislike it - please "Unresolve" this conversation so I will remove it :) and please feel free to do not spend time on explanation if you lack time (although would be just interesting and higly appreciated for me personally to hear your point of view on this here just to get some lesson/knowledge exchange here).